File: sort/counting_sort.cpp
#include <vector>
#include <iostream>

using namespace std;

void countingSort(vector<int>& nums) {
    int max = *max_element(nums.begin(), nums.end());
    vector<int> counter(max + 1);
    for (int i = 0; i < nums.size(); i++) {
        counter[nums[i]]++;
    }
    vector<int> result(nums.size());
    nums = {};
    for (int i = 0; i < counter.size(); i++) {
        while(static_cast<bool>(counter[i]--)) {
            nums.push_back(i);
        }
    }
}

int main(int argc, char const *argv[])
{
    vector<int> nums = {10, 3, 1, 8, 4, 7, 6};
    cout << "Before Sorted: ";
    for (int num : nums) {
        cout << num << " ";
    }
    cout << endl;
    
    countingSort(nums);
    cout << "After Sorted: ";
    for (int num : nums) {
        cout << num << " ";
    }
    cout << endl;
    return 0;
}

File: sort/counting_sort.py
def countingSort(nums: list):
    maxValue = max(nums)
    counter = [0] * (maxValue + 1)
    for num in nums:
        counter[num] += 1
    nums.clear()
    nums = [] # 定义局部变量会在函数结束时摧毁，不会传递到入参的nums
    for i in range(0, len(counter)):
        while (counter[i]):
            nums.append(i)
            counter[i] -= 1
    
if __name__ == "__main__":
    nums = [10, 3, 1, 8, 4, 7, 6]
    print(f"Before Sorted: {nums}")
    countingSort(nums)
    print(f"After Sorted: {nums}")